system-contracts/test/L2EthToken.spec.ts

```diff
@@ -0,0 +1,222 @@
+    });
+  });
+
+  describe("name", () => {
+    it("correct name", async () => {
+      const name = await l2EthToken.name();
+      expect(name).to.equal("Ether");
+    });
+  });
+
+  describe("symbol", () => {
+    it("correct symbol", async () => {
+      const symbol = await l2EthToken.symbol();
+      expect(symbol).to.equal("ETH");
+    });
+  });
+
+  describe("decimals", () => {
+    it("correct decimals", async () => {
+      const decimals = await l2EthToken.decimals();
+      expect(decimals).to.equal(18);
+    });
+  });
+
+  describe("withdraw", () => {
+    it("event, balance, totalsupply", async () => {
+      const selector = IMailboxFactory.connect("0x00", provider).interface.getSighash("finalizeEthWithdrawal");
+      const amountToWithdraw: BigNumber = ethers.utils.parseEther("1.0");
+
+      const message: string = ethers.utils.solidityPack(
+        ["bytes4", "address", "uint256"],
+        [selector, wallets[1].address, amountToWithdraw]
+      );
+
+      await setResult("L1Messenger", "sendToL1", [message], {
+        failure: false,
+        returnData: ethers.utils.defaultAbiCoder.encode(["bytes32"], [ethers.utils.keccak256(message)]),
+      });
+
+      const amountToMint: BigNumber = ethers.utils.parseEther("100.0");
+      await (await l2EthToken.connect(bootloaderAccount).mint(l2EthToken.address, amountToMint)).wait();
+
+      const balanceBeforeWithdrawal: BigNumber = await l2EthToken.balanceOf(l2EthToken.address);
+      const totalSupplyBefore = await l2EthToken.totalSupply();
+
+      await expect(l2EthToken.connect(richWallet).withdraw(wallets[1].address, { value: amountToWithdraw }))
+        .to.emit(l2EthToken, "Withdrawal")
+        .withArgs(richWallet.address, wallets[1].address, amountToWithdraw);
+
+      const balanceAfterWithdrawal: BigNumber = await l2EthToken.balanceOf(l2EthToken.address);
+      const totalSupplyAfter = await l2EthToken.totalSupply();
+
+      expect(balanceAfterWithdrawal).to.equal(balanceBeforeWithdrawal.sub(amountToWithdraw));
+      expect(totalSupplyAfter).to.equal(totalSupplyBefore.sub(amountToWithdraw));
+    });
+
+    it("event, balance, totalsupply, withdrawWithMessage", async () => {
+      const selector = IMailboxFactory.connect("0x00", provider).interface.getSighash("finalizeEthWithdrawal");
+      const amountToWidthdraw: BigNumber = ethers.utils.parseEther("1.0");
+      const additionalData: string = ethers.utils.defaultAbiCoder.encode(["string"], ["additional data"]);
+      const message: string = ethers.utils.solidityPack(
+        ["bytes4", "address", "uint256", "address", "bytes"],
+        [selector, wallets[1].address, amountToWidthdraw, richWallet.address, additionalData]
+      );
+
+      await setResult("L1Messenger", "sendToL1", [message], {
+        failure: false,
+        returnData: ethers.utils.defaultAbiCoder.encode(["bytes32"], [ethers.utils.keccak256(message)]),
+      });
+
```

Comment:
Can you also add mint on the `l2EthToken` address here or make this mint in some `before` block? I understand that in the current test order it works, but it would be better for consistency.